competition-simulator.md

```diff
@@ -0,0 +1,120 @@
+---
+layout: page
+title: Competition Simulator
+---
+
+Competition Simulator
+============
+
+For SR2020, a virtual competition will be run using a brand new simulator.
+
+<div class="info">
+  This is different to the <a href="/docs/programming/simulator/">regular simulator</a>.
+</div>
+
+## Installation
+
+### Prerequisites
+
+This simulation is based in [Webots](https://cyberbotics.com/#download), which will need to be downloaded and installed.
+
+You will also need Python installed. The simulator supports >=3.4.
+
+### Installing the simulation
+
+1. [Download the simulation](), and unzip it somewhere on your computer.
+2. Using the Webots IDE, open the `worlds/Arena.wbt` file
+
+## Overview
+
+Within the IDE, there are a few different panels:
+
+- In the centre of your screen is the 3D simulated view of the arena
+- On the left is a tree hierarchy of all elements in this "world"
+- On the right is your text editor for modifying the `sr_controller` code for example
+- At the bottom is your console
+- At the top are your general controls which include the time controls. Press the centre play button to run the simulation at normal speed.
+
+**Important:** Changes to the world must happen with the simulation paused at 0:00. If e.g. you move an object at a different time, rewinding back to the start will delete your changes.
+
+### Useful links
+
+- [Camera Controls](https://www.cyberbotics.com/doc/guide/the-3d-window#navigation-in-the-scene)
+- [Graphics settings](https://www.cyberbotics.com/doc/guide/preferences#opengl) (Useful if Webots is running slowly)
+
+## Developing your code
+
+The code for your robot is `controllers/sr_controller/sr_controller.py`. An example skeleton code is provided for convenience.
+
+<div class="warning">
+  Please refrain from editing any code in the <code>sr</code> module directory. This contains the code which maps the SR API onto the webots API.
+  Any changes you make to this directory will not be present in the competition environment.
+</div>
+
+### Updates
+
+Occasionally, we may release an update to the simulation. To update, you will need to copy your code our of of the `competition-simulator-master` directory, delete the directory, and re-download it using the above link. Then you can replace the new `controllers/sr_controller/sr_controller.py` with your existing code.
```

Comment:
Hrm, this approach feels quite likely to result in "oops I accidentally deleted our code" from people just blatting the new directory on top of the old one.

Could we do something to avoid that? For example, can we arrange it so that the competitors code is meant to live _outside_ the directory they downloaded?

Related to this: does Webots come with a Python IDE, or are they going to use something of their own? If the former, would moving the code outside this structure complicate that?

Reply:
Webots does have an editor, and yes moving the code outside would complicate things. I've intentionally written things such that they can use their own editors.

Moving the code primarily outside would be possible, and we could probably do that with a custom controller.

Alternatively, we should encourage regular backups of code, and good version control practice, regardless of this update mechanism

Reply:
I'm tempted to suggest they fork the git repo for their development. (Although this doesn't seem like the perfect solution)

Reply:
That's definitely not a viable option, as many competitors won't have the slightest idea what `git` is, let alone how to use it properly.

Reply:
Them using forks makes updates much harder (merge conflicts also being possible) as well as meaning that we'd effectively need to support every commit, rather than a small number of blessed releases.

Reply:
@RealOrangeOne I think we should consider this blocking for release -- I don't think we're going to get a second chance to change how teams set up their development environments.

I suggest we fix this by having our `sr_controller.py` look for their code in some known location, perhaps a sibling `robot` directory to the place the simulator is checked out to.

From my testing, the built-in editor lets you open files anywhere you like and didn't seem to have any special affinity for the files within itself anyway.

If we're going to encourage that they use other editors (which I think we should - that editor is pretty bad), then we might as well have the code somewhere they can feel more in control of it. There might be a bit more setup for their editors, but I think that that's worth the trade-off of not losing code (especially since the IDE made it _hard_ to lose code, but this current mechanism makes it a loaded foot-gun).